# Data loading, cleanup and processing

The first step to a ML project is to obtain the dataset you will be working with. 
There are many repositories for materials science-specific data (whether online or offline)---consult the accompanying paper for a list of the more commonly used ones.

Once you have identified the repository and dataset you will use for your project, you will have to download it to your local machine, or establish a way to reliably access the dataset.
Consult the documentation of the repository for how to do this.

In [5]:
import os
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline
%config InlineBackend.figure_format='retina'

#from pandas_profiling import ProfileReport

## Load data

Using Pandas, we read in the dataset into a DataFrame. 

We also print the shape of the DataFrame, which indicates the number of rows and columns in this dataset.

In [6]:
PATH = os.getcwd()
data_path = os.path.join(PATH, '../data_abs/cp_data_abs_demo.csv')

df = pd.read_csv(data_path)
print(f'Original DataFrame shape: {df.shape}')

Original DataFrame shape: (3612, 3)


This means that our input dataset has 3612 data samples, each with 3 variables.

## Examine the data

We examine some rows and look at the data's basic statistics.

We see that the dataset contains information about the formula, corresponding energy (in this case, temperature in K), and the target property, absorption coefficient.

We can get some simple summary statistics of the DataFrame by calling the `.describe()` method on the database.

In [5]:
df.describe()

energy          abs
count  3612.000000  3612.000000
mean      2.500000     0.224144
std       1.448379     0.554535
min       0.000000     0.000000
25%       1.250000     0.010410
50%       2.500000     0.023550
75%       3.750000     0.098597
max       5.000000     3.196810

In [13]:
df.shape

(3612, 3)

## Rename the column names for brevity

In [6]:
df.columns

Index(['formula', 'energy', 'abs'], dtype='object')

In [7]:
#rename_dict = {'FORMULA': 'formula',
               #'CONDITION: : '',
               #'PROPERTY: : ''}
#df = df.rename(columns=rename_dict)
df.columns

Index(['formula', 'energy', 'abs'], dtype='object')

## Check for and remove `NaN` values

Here we can use the built-in Pandas methods to check for `NaN` values in the dataset, which are missing values.
We then remove the dataset rows which contain `NaN` values.

In [8]:
# Check for NaNs in the respective dataset columns, and get the indices
df2 = df.copy()
bool_nans_formula = df2['formula'].isnull()
bool_nans_E = df2['energy'].isnull()
bool_nans_Abs = df2['abs'].isnull()

# Drop the rows of the DataFrame which contain NaNs
df2 = df2.drop(df2.loc[bool_nans_formula].index, axis=0)
df2 = df2.drop(df2.loc[bool_nans_E].index, axis=0)
df2 = df2.drop(df2.loc[bool_nans_Abs].index, axis=0)

print(f'DataFrame shape before dropping NaNs: {df.shape}')
print(f'DataFrame shape after dropping NaNs: {df2.shape}')

DataFrame shape before dropping NaNs: (3612, 3)
DataFrame shape after dropping NaNs: (3612, 3)


In [9]:
df3 = df.copy()
df3 = df3.dropna(axis=0, how='any')

print(f'DataFrame shape before dropping NaNs: {df.shape}')
print(f'DataFrame shape after dropping NaNs: {df3.shape}')

df = df3.copy()

DataFrame shape before dropping NaNs: (3612, 3)
DataFrame shape after dropping NaNs: (3612, 3)


## Check for and remove unrealistic values

In some cases, you might also get data values that simply don't make sense.
For our dase, this could be negative values in the temperature or heat capacity values.

In [10]:
bool_invalid_E = df['energy'] < 0
bool_invalid_Abs = df['abs'] < 0

df = df.drop(df.loc[bool_invalid_E].index, axis=0)
df = df.drop(df.loc[bool_invalid_Abs].index, axis=0)

print(f'DataFrame shape after dropping unrealistic values: {df.shape}')

DataFrame shape after dropping unrealistic values: (3612, 3)


## Save cleaned data to csv

Finally, after cleaning and processing the data, you can save it to disk in a cleaned state for you to use later.

Pandas allows us to save our data as a comma separated value `.csv` file. 

In [12]:
out_path = os.path.join(PATH, '../data_abs/cp_data_abs_cleaned.csv')
df.to_csv(out_path, index=False)

Note, your data can be saved in other file formats (such as hdf5) or in databases (such as SQL), but we will not go into the details of these formats.

Typically, the amount of data you can gather for your ML project isn't large enough to warrant these approaches.